pinot-core/src/main/java/org/apache/pinot/core/data/table/TableResizer.java

```diff
@@ -136,39 +138,46 @@ private IntermediateRecord getIntermediateRecord(Key key, Record record) {
    * Resize only if number of records is greater than trimToSize
    * The resizer smartly chooses to create PQ of records to evict or records to retain, based on the number of records and the number of records to evict
    */
-  public void resizeRecordsMap(Map<Key, Record> recordsMap, int trimToSize) {
+  public Map<Key, Record> resizeRecordsMap(Map<Key, Record> recordsMap, int trimToSize) {
     int numRecordsToEvict = recordsMap.size() - trimToSize;
-
     if (numRecordsToEvict > 0) {
       // TODO: compare the performance of converting to IntermediateRecord vs keeping Record, in cases where we do not need to extract final results
-
-      if (numRecordsToEvict < trimToSize) { // num records to evict is smaller than num records to retain
+      if (numRecordsToEvict < trimToSize) {
+        // num records to evict is smaller than num records to retain
         // make PQ of records to evict
         PriorityQueue<IntermediateRecord> priorityQueue =
             convertToIntermediateRecordsPQ(recordsMap, numRecordsToEvict, _intermediateRecordComparator);
         for (IntermediateRecord evictRecord : priorityQueue) {
           recordsMap.remove(evictRecord._key);
         }
-      } else { // num records to retain is smaller than num records to evict
+        return recordsMap;
+      } else {
+        // num records to retain is smaller than num records to evict
         // make PQ of records to retain
+        Map<Key, Record> trimmedRecordsMap;
```

Comment:
Try to not create a new map as growing the map again is very expensive

Reply:
@Jackie-Jiang , I noticed that retainAll() is very expensive so creating a result map (same as what is done by AggregationGroupByTrimmingService in PQL) is preferrable and it resulted in significant improvement. retainAll() was taking > 1sec in some cases

Reply:
We need to avoid `retainAll()` on PQ for sure, but we should also try to reuse the existing map